### ISSUE TITLE
Unable to reproduce BERT results.

### ISSUE THREAD

**Issue by Haihua-Luo** [Thartvigsen/GRACE#10]
Hello, I'm trying to reproduce the main result of **BERT** in Table 2, but I obtained different results. The main result in your paper is TRR=0.81, ERR=0.82, #E=381, while I get TRR=0.97, ERR=1, #E=340. Is there any issue with my configuration files?

Here is my log file:

```bash
CUDA_VISIBLE_DEVICES=0 python grace/main.py experiment=scotus model=scotus-bert editor=grace
[2024-12-03 05:23:59,353][__main__][INFO] -

tags: null
batch_size: 1
n_iter: 100
max_n_edits: 1000
reinit: true
dropout: 0.0
device: cuda
wandb: false
wandb_mode: online
wandb_project_name: my_project
wandb_run_name: null
metric_period: 50
pretrain: false
load_pretrained: true
re_init_model: false
ckpt: false
ckpt_dir: ./ckpts/
editor:
_name: grace
edit_lr: 1.0
n_iter: 100
eps: 1.0
dist_fn: euc
val_init: cold
val_train: sgd
val_reg: None
reg: early_stop
replacement: replace_prompt
eps_expand: coverage
num_pert: 8
experiment:
task: scotus
model:
name: tomh/scotus-bert
class_name: AutoModelForSequenceClassification
tokenizer_class: AutoTokenizer
tokenizer_name: ./grace/model/scotus-bert
inner_params:

- bert.encoder.layer[10].output.dense.weight
  pt: ./grace/model/scotus-bert

[2024-12-03 05:23:59,353][__main__][INFO] - Project base directory: /home/ranxuming/project/haihua/GRACE-main
{'tags': None, 'batch_size': 1, 'n_iter': 100, 'max_n_edits': 1000, 'reinit': True, 'dropout': 0.0, 'device': 'cuda', 'wandb': False, 'wandb_mode': 'online', 'wandb_project_name': 'my_project',
'wandb_run_name': None, 'metric_period': 50, 'pretrain': False, 'load_pretrained': True, 're_init_model': False, 'ckpt': False, 'ckpt_dir': './ckpts/',
'editor': {'_name': 'grace', 'edit_lr': 1.0, 'n_iter': 100, 'eps': 1.0, 'dist_fn': 'euc', 'val_init': 'cold', 'val_train': 'sgd', 'val_reg': 'None', 'reg': 'early_stop', 'replacement': 'replace_prompt',
'eps_expand': 'coverage', 'num_pert': 8}, 'experiment': {'task': 'scotus'}, 'model': {'name': 'tomh/scotus-bert', 'class_name': 'AutoModelForSequenceClassification', 'tokenizer_class':
'AutoTokenizer', 'tokenizer_name': './grace/model/scotus-bert', 'inner_params': ['bert.encoder.layer[10].output.dense.weight'], 'pt': './grace/model/scotus-bert'}}
[2024-12-03 05:23:59,450][grace.models][INFO] - Loading model class <class 'transformers.models.auto.modeling_auto.AutoModelForSequenceClassification'>
with name tomh/scotus-bert from cache dir ./grace/model/scotus-bert
Loaded model from local directory: ./grace/model/scotus-bert
Loaded 931 candidate edits.
Loaded editor: grace.
Editor inherited .generate.

Average performance on edit set:  tensor(0.5542)
Original TRR:  tensor(0.9879)
0it [00:00, ?it/s]0
tensor([True], device='cuda:0')

Number of edits: 1
TRR: 0.9878706336021423
ERR: 1.0
ES: 1.0
train_time: 0.023205494880676268
edit_label: tensor([2])
n_edits: 1
chosen_key: tensor([0], device='cuda:0')
nkeys: 1
50it [00:34,  1.23it/s]50
tensor([True], device='cuda:0')

Number of edits: 25
TRR: 0.9878706336021423
ERR: 1.0
ES: 1.0
train_time: 0.021669387817382812
edit_label: tensor([7])
n_edits: 25
chosen_key: tensor([24], device='cuda:0')
nkeys: 25
98it [01:46,  3.98s/it]100
tensor([True], device='cuda:0')

Number of edits: 53
TRR: 0.9824798107147217
ERR: 1.0
ES: 1.0
train_time: 0.022392412026723225
edit_label: tensor([7])
n_edits: 53
chosen_key: tensor([52], device='cuda:0')
nkeys: 53
149it [02:25,  1.43it/s]150
tensor([True], device='cuda:0')

Number of edits: 80
TRR: 0.9824798107147217
ERR: 1.0
ES: 1.0
train_time: 0.02253512144088745
edit_label: tensor([7])
n_edits: 80
chosen_key: tensor([79], device='cuda:0')
nkeys: 80
200it [03:29,  4.05s/it]200
tensor([False], device='cuda:0')
247it [04:01,  2.07it/s]250
tensor([True], device='cuda:0')

Number of edits: 124
TRR: 0.979784369468689
ERR: 1.0
ES: 1.0
train_time: 0.022911429405212402
edit_label: tensor([7])
n_edits: 124
chosen_key: tensor([123], device='cuda:0')
nkeys: 124
300it [04:29,  2.08it/s]300
tensor([True], device='cuda:0')

Number of edits: 143
TRR: 0.979784369468689
ERR: 1.0
ES: 1.0
train_time: 0.022100396951039634
edit_label: tensor([7])
n_edits: 143
chosen_key: tensor([142], device='cuda:0')
nkeys: 143
349it [05:31,  2.78it/s]350
tensor([False], device='cuda:0')
398it [06:01,  1.69it/s]400
tensor([False], device='cuda:0')
450it [07:13,  1.42s/it]450
tensor([False], device='cuda:0')
500it [07:41,  2.79it/s]500
tensor([False], device='cuda:0')
550it [08:51,  4.96s/it]550
tensor([False], device='cuda:0')
599it [09:20,  1.01it/s]600
tensor([True], device='cuda:0')

Number of edits: 271
TRR: 0.9743935465812683
ERR: 1.0
ES: 1.0
train_time: 0.02237794001897176
edit_label: tensor([0])
n_edits: 271
chosen_key: tensor([269], device='cuda:0')
nkeys: 270
650it [09:52,  2.59it/s]650
tensor([True], device='cuda:0')

Number of edits: 293
TRR: 0.9743935465812683
ERR: 1.0
ES: 1.0
train_time: 0.023625886440277098
edit_label: tensor([7])
n_edits: 293
chosen_key: tensor([291], device='cuda:0')
nkeys: 292
700it [11:03,  1.38it/s]700
tensor([True], device='cuda:0')

Number of edits: 317
TRR: 0.973045825958252
ERR: 1.0
ES: 1.0
train_time: 0.022406172752380372
edit_label: tensor([9])
n_edits: 317
chosen_key: tensor([315], device='cuda:0')
nkeys: 316
750it [11:37,  1.14it/s]750
tensor([True], device='cuda:0')

Number of edits: 340
TRR: 0.973045825958252
ERR: 1.0
ES: 1.0
train_time: 0.02272079388300578
edit_label: tensor([3])
n_edits: 340
chosen_key: tensor([336], device='cuda:0')
nkeys: 337
798it [12:47,  1.25it/s]800
tensor([False], device='cuda:0')
848it [13:12,  2.44it/s]850
tensor([False], device='cuda:0')
898it [13:38,  1.89it/s]900
tensor([False], device='cuda:0')
931it [14:32,  1.07it/s]
```

**Comment by Haihua-Luo** [comment]
I want to add more details

I download the pre-trained **BERT** to local and upload to serve because my serve couldn't connect to Huggingface.

Here is my download code:

```Bash
model_name = "tomh/scotus-bert"
model_path = "D:/Code/GRACE-main/hugg/scotus-bert"  

model = AutoModelForSequenceClassification.from_pretrained(model_name)
model.save_pretrained(model_path)

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")
tokenizer.save_pretrained(model_path)
```

and then i upload it to my serve at ./grace/model/scotus-bert and change the way of loading model to load the BERT in **./grace/model/scotus-bert** directly

I also download the datasets
```bash
train_data = load_dataset("tomh/grace-scotus", split="train")
test_data = load_dataset("tomh/grace-scotus", split="test")

save_path = "D:/LHH/research/Code/GRACE-main/hugg/scotus_dataset"

train_data.to_csv(os.path.join(save_path, "train.csv"))
test_data.to_csv(os.path.join(save_path, "test.csv"))

# convert CSV file to Arrow file
dataset_path = "./hugg/scotus_dataset"
train_df = pd.read_csv(os.path.join(dataset_path, "train.csv"))
test_df = pd.read_csv(os.path.join(dataset_path, "test.csv"))

train_dataset = Dataset.from_pandas(train_df)
test_dataset = Dataset.from_pandas(test_df)

train_dataset.save_to_disk(os.path.join(dataset_path, "train.arrow"))
test_dataset.save_to_disk(os.path.join(dataset_path, "test.arrow"))
```

also upload it to serve in **./grace/data/scotus_dataset**

**Comment by Thartvigsen** [comment]
Thanks for your interest in our work! I sadly didn't add the hparam settings for each experiment when originally uploading the repo and I sadly don't have the time to dig through the logs from the original experiments. But I bet this is a hyperparameter issue. So the first thing I suggest is to try different values of ```replacement``` in the config. If I remember correctly, it should be ```replacement: replace_all``` for BERT experiments. But you might also try ```replacement: replace_last```. Sorry for the hassle, but I'm curious how much of an impact this has! It's important to note that in your experiment, the number of keys is almost equal to the number of edits, so it's really just memorizing every edit. You might also try ```val_init: warm``` if ```replacement``` doesn't fix it.

**Comment by Haihua-Luo** [comment]
Thank you for your response.

Is this happen because i download the pre_trained BERT? i notice that there are words in paper appendix B.2

>We finetune a 110 million parameter BERT model (bert-base-cased on Huggingface) on
the SCOTUS training using Huggingface. We use 50 epochs with a batch size of 4 and a learning rate
of 5e−05 with an AdamW optimizer using Huggingface’s default model trainer. The finetuned BERT
model achieves 0.99 Accuracy on its original validation data and 0.55 on the edit set, indicating a
large amount of improvement to be made from editing.

this is my  result from pre-trained Bert now:

```
[2024-12-03 05:23:59,353][__main__][INFO] - 

tags: null
batch_size: 1
n_iter: 100
max_n_edits: 1000
reinit: true
dropout: 0.0
device: cuda
wandb: false
wandb_mode: online
wandb_project_name: my_project
wandb_run_name: null
metric_period: 50
pretrain: false
load_pretrained: true
re_init_model: false
ckpt: false
ckpt_dir: ./ckpts/
editor:
  _name: grace
  edit_lr: 1.0
  n_iter: 100
  eps: 1.0
  dist_fn: euc
  val_init: cold
  val_train: sgd
  val_reg: None
  reg: early_stop
  replacement: replace_prompt
  eps_expand: coverage
  num_pert: 8
experiment:
  task: scotus
model:
  name: tomh/scotus-bert
  class_name: AutoModelForSequenceClassification
  tokenizer_class: AutoTokenizer
  tokenizer_name: ./grace/model/scotus-bert
  inner_params:
  - bert.encoder.layer[10].output.dense.weight
  pt: ./grace/model/scotus-bert

[2024-12-03 05:23:59,450][grace.models][INFO] - Loading model class <class 'transformers.models.auto.modeling_auto.AutoModelForSequenceClassification'> 
with name tomh/scotus-bert from cache dir ./grace/model/scotus-bert
Loaded model from local directory: ./grace/model/scotus-bert
Loaded 931 candidate edits.
Loaded editor: grace.
Editor inherited .generate.


Average performance on edit set:  tensor(0.5542)
Original TRR:  tensor(0.9879)
```

Need i use the finetuned Bert? and how can i get it?

**Comment by Thartvigsen** [comment]
Are you using our trained BERT? ```https://huggingface.co/tomh/scotus-bert```

**Comment by Haihua-Luo** [comment]
yes, i download model form huggingface:

```
# download BERT model
model_name = "tomh/scotus-bert"

# download model and save to local
model = AutoModelForSequenceClassification.from_pretrained(model_name)
model_path = "D:/research/Code/GRACE-main/hugg/scotus-bert"
model.save_pretrained(model_path)

# use bert-base-cased as tokenizer
tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")
tokenizer.save_pretrained(model_path)
```

and then i upload the downloaded model file form local to serve, so i get the original result:
```
Average performance on edit set:  tensor(0.5542)
Original TRR:  tensor(0.9879)
```

**Comment by Thartvigsen** [comment]
Ok perfect! Let me know the results of my suggestions above for hyperparameters. Otherwise, it could be a seed issue.

**Comment by Haihua-Luo** [comment]
this is the hyperparameter of GPT2-xl and Bert with seed 42, still different, what should i do next?

![1](https://github.com/user-attachments/assets/5108a5aa-bdad-4f7c-b8e7-30a60fc73a56)

![2](https://github.com/user-attachments/assets/24adbf0d-9913-4267-a457-c810f2e476f8)

**Comment by Thartvigsen** [comment]
It looks like increasing the epsilon gets very close to the numbers reported in the paper. It's possible that we did not control the seed in this experiment in the original submission.

**Comment by Haihua-Luo** [comment]
Yes, it's closer, but still a little difference, i will try other hyperparameter, thank you for your response.

If you find your experiment settings or any further suggestions, please tell me.
